galley/pkg/util/worker.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+	"sync"
+
+	"istio.io/istio/pkg/log"
+)
+
+// Worker is a utility to help manage the lifecycle of a worker thread.
+type Worker struct {
+	// name of the worker thread used for logging.
+	name string
+
+	// scope is the logging scope to be used for log messages.
+	scope log.Scope
+
+	// Lock for changing the running state of the source
+	stateLock sync.Mutex
+
+	// started indicates that the user has started this source (via Start).
+	started bool
+
+	// stopped indicates that the user has stopped this source (via Stop).
+	stopped bool
+
+	// stopCh is used to trigger the k8s informer to stop.
+	stopCh chan struct{}
+
+	// stoppedCh is used to indicate that the k8s informer has exited.
+	stoppedCh chan struct{}
+}
+
+// NewWorker creates a new worker with the given name and logging scope.
+func NewWorker(name string, scope *log.Scope) *Worker {
+	if scope == nil {
+		scope = log.FindScope(log.DefaultScopeName)
+	}
+	return &Worker{
+		name:      name,
+		scope:     *scope,
+		stopCh:    make(chan struct{}, 1),
+		stoppedCh: make(chan struct{}, 1),
+	}
+}
+
+// Start the worker thread via the provided function. The startWorkerThread function is
+// provided two control channels. When the user closes this worker, stopCh is closed to
+// signal that the worker thread should begin shutting down. Upon exiting, the worker
+// thread should in turn close stoppedCh to signal back to the worker that the thread
+// has terminated.
+func (w *Worker) Start(startWorkerThread func(stopCh chan struct{}, stoppedCh chan struct{}) error) error {
+	w.stateLock.Lock()
+	defer w.stateLock.Unlock()
+
+	if w.started {
+		return fmt.Errorf("%s already started", w.name)
+	}
+
+	err := startWorkerThread(w.stopCh, w.stoppedCh)
+	if err == nil {
+		w.started = true
+	}
+	return err
+}
+
+// Stop the worker thread. Once the worker thread has terminated, the provided cleanup
+// function is executed.
+func (w *Worker) Stop(preStopFn func(), postStopFn func()) {
```

Comment:
As mentioned above, I think pre/post methods are making it hard to understand this code.  The only value add of executing these methods within this context is to execute them under Worker's lock, which is subtle and hard to reason about. IMO, these should be handled as part of client's code, and it should be upto the client to properly synchronize them.

For most cases, this should be trivial, as a preStopFn should just execute as last part of the worker go routine, so it can be embedded into the worker lambda. postStopFn can be executed after stop returns.

Reply:
I think it will also reduce the booleans (started/stopped). You should be able to get away with using a single boolean (or existence of a member).

Reply:
done.